packages/flutter/lib/src/services/message_codecs.dart

```diff
@@ -386,9 +386,28 @@ class StandardMessageCodec implements MessageCodec<Object?> {
       }
     } else if (value is String) {
       buffer.putUint8(_valueString);
-      final Uint8List bytes = utf8.encoder.convert(value);
-      writeSize(buffer, bytes.length);
-      buffer.putUint8List(bytes);
+      final Uint8List asciiBytes = Uint8List(value.length);
+      Uint8List? utf8Bytes;
+      int utf8Offset = 0;
+      // Only do utf8 encoding if we encounter non-ascii characters.
```

Comment:
I though dart utf8 encoding already had a fast path for this?

Reply:
Yep, but here are the reasons I believe we are getting faster results:
1) We are removing an extra copy of the data: https://github.com/dart-lang/sdk/blob/56035a7df0bb26a6babce53fae21d46263f3bf26/sdk/lib/convert/utf.dart#L115
1) It inlines the logic
1) We are removing bounds checks https://github.com/dart-lang/sdk/blob/56035a7df0bb26a6babce53fae21d46263f3bf26/sdk/lib/convert/utf.dart#L98

Reply:
Here's another bounds check we were able to remove: https://github.com/dart-lang/sdk/blob/56035a7df0bb26a6babce53fae21d46263f3bf26/sdk/lib/convert/utf.dart#L201

Reply:
Dart's UTF8 encoder main loop is: https://github.com/dart-lang/sdk/blob/main/sdk/lib/convert/utf.dart#L197

For strings containing only ASCII Dart should just take each code unit and copy it into the output byte buffer, which is similar to what this loop is doing.  I'm not sure why there would be a noticeable performance difference between Dart's encoder and this PR.

Reply:
Ahh I see. Darts encoder guarantees that the Uint8List it gives you is not just a view of some larger buffer somewhere else, but we don't need to worry about that since we're immediately writing this into another one.

Makes sense!

Reply:
That makes sense - the typed data `sublist` is doing a memcpy.  But this encoder can avoid that by writing the ASCII part and the non-ASCII part to the `WriteBuffer` as two separate chunks.